internal/repository/attestations.go

```diff
@@ -255,6 +267,188 @@ func (r *Repository) AddGitHubPullRequestAttestationForNumber(ctx context.Contex
+		if err := json.Unmarshal(payloadBytes, stmt); err != nil {
+			return err
+		}
+
+		approvers = append(approvers, stmt.Predicate.Approvers...)
+		dismissedApprovers = stmt.Predicate.DismissedApprovers
+	}
+
+	approvers = getFilteredSetOfApprovers(approvers)
+	dismissedApprovers = getFilteredSetOfApprovers(dismissedApprovers)
+
+	statement, err := attestations.NewGitHubPullRequestApprovalAttestation(baseRef, fromID, toID, approvers, dismissedApprovers)
+	if err != nil {
+		return err
+	}
+
+	env, err = dsse.CreateEnvelope(statement)
+	if err != nil {
+		return err
+	}
+
+	keyID, err := signer.KeyID()
+	if err != nil {
+		return err
+	}
+
+	slog.Debug(fmt.Sprintf("Signing GitHub pull request approval attestation using '%s'...", keyID))
+	env, err = dsse.SignEnvelope(ctx, env, signer)
+	if err != nil {
+		return err
+	}
+
+	if err := currentAttestations.SetGitHubPullRequestApprovalAttestation(r.r, env, reviewID, baseRef, fromID, toID); err != nil {
+		return err
+	}
+
+	commitMessage := fmt.Sprintf("Add GitHub pull request approval for '%s' from '%s' to '%s' (review ID %d) for approval by '%s'", baseRef, fromID, toID, reviewID, approver.KeyID)
+
+	slog.Debug("Committing attestations...")
+	return currentAttestations.Commit(r.r, commitMessage, signCommit)
+}
+
+func (r *Repository) DismissGitHubPullRequestApprover(ctx context.Context, signer sslibdsse.SignerVerifier, reviewID int64, dismissedApprover *tuf.Key, signCommit bool) error {
+	if !dev.InDevMode() {
+		return dev.ErrNotInDevMode
+	}
+
+	currentAttestations, err := attestations.LoadCurrentAttestations(r.r)
+	if err != nil {
+		return err
+	}
+	indexPath, has := currentAttestations.GetGitHubPullRequestApprovalIndexPathForReviewID(reviewID)
+	if !has {
+		return attestations.ErrGitHubReviewIDNotFound
+	}
+
+	// TODO: if the helper above has an indexPath, we can directly load that blob, simplifying the logic here
+	hasApprovalAttestation := false
+	env, err := currentAttestations.GetGitHubPullRequestApprovalAttestationForIndexPath(r.r, indexPath)
+	if err != nil {
+		return err
+	}
+
+	dismissedApprovers := []*tuf.Key{dismissedApprover}
+	var approvers []*tuf.Key
+	if !hasApprovalAttestation {
+		// Create a new GitHub pull request approval attestation
+		slog.Debug("Creating new GitHub pull request approval attestation...")
+	} else {
+		// Update existing statement's predicate and create new env
+		slog.Debug("Adding approver to existing GitHub pull request approval attestation...")
+		payloadBytes, err := env.DecodeB64Payload()
+		if err != nil {
+			return err
+		}
+
+		type tmpStatement struct {
+			Type          string                                             `json:"_type"`
+			Subject       []*ita.ResourceDescriptor                          `json:"subject"`
+			PredicateType string                                             `json:"predicateType"`
+			Predicate     *attestations.GitHubPullRequestApprovalAttestation `json:"predicate"`
+		}
```

Comment:
If you're defining this multiple times, you're probably better off moving this out into it's own type.

Reply:
In general, this looks largely similar to AddGitHubPullRequestApprover - can we refactor this to move common parsing/writing to a shared function?

Reply:
:eyes: